torch/_dynamo/variables/misc.py

```diff
@@ -860,7 +860,32 @@ def __init__(self, value, **kwargs):
     def call_function(
         self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
     ) -> "VariableTracker":
-        unimplemented("numpy")
+        if not config.numpy_ndarray_as_tensor or not HAS_NUMPY_TORCH_INTEROP:
+            unimplemented(f"numpy.{self.value}()")
+        import torch_np._detail.implementations
+
+        from .builder import wrap_fx_proxy_cls
+        from .tensor import NumpyTensorVariable
+
+        options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
+        # lookup method name in torch_np
+        if self.value.__name__ in torch_np._detail.implementations.__dict__:
```

Comment:
@rgommers is this the correct way to look up the implementation of a function from torch_np? This... doesn't seem quite right to me, IMO.

Reply:
hmm would it be better if I change it to inspect.getattr_static

Reply:
So, I guess what I was expecting, is either `torch_np` has some mapping of name to function, so that you just do a lookup from the mapping, or there is some convention that `torch_np.XXX` corresponds to `numpy.XXX`, so you do a getattr on the module itself (with possibly some handling for numpy.submodule)

Reply:
> or there is some convention that `torch_np.XXX` corresponds to `numpy.XXX`,

that should be the case

Reply:
So I can't just use `torch_np.meshgrid` since it's mapping numpy.meshgrid return (a list of ndarray)

Reply:
using `torch_np._details` instead

Reply:
> So I can't just use torch_np.meshgrid since it's mapping numpy.meshgrid return (a list of ndarray)

Can you elaborate on that? The `torch_np` implementations and the `torch_np.ndarray` object are designed to be drop-in replacements for `np` functions. You should be able to use them here (in fact, that's their only reason of existence). Even more, you should not use things from the repo with an underscore, as those are meant to be implementation details.

Reply:
@lezcano: `torch_np.meshgrid` currently takes in 2 torch tensors but returns a list of np ndarrays. Directly using that doesn't fit what we want - we want torch tensor in torch tensor out. Instead `torch_np._detail.implementations.meshgrid` is what we want.

Reply:
That's not quite right. All the functions in `torch_np` accept `torch_np.ndarray`s (and other array-like things that NumPy accepts like lists of lists or simply python or numpy scalars) and return a `torch_np.ndarray`. `torch_np.ndarray` is just a thin wrapper around `torch.Tensor`, so here, you should simply access `ndarray.tensor` and use that internally.

Reply:
@lezcano hmm it's not simple as that, I understand we need to pass `torch_np.ndarray` around for `torch_np.meshgrid` and other APIs, the thing is on dynamo (AFAIK) we can't support arbitrary class, even it's just a simple wrapper of torch.Tensor. 

My understanding for how dynamo works is that it when we call `np.meshgrid`, we will need to translate that to some other function, but only taking torch.Tensor as input and output. Now if I use `torch_np.meshgrid` I can't just get back the result and access `ndarray.tensor` in `NumpyTensorVariable`, since we don't actually run `torch_np.meshgrid` during dynamo compilation and hence no result will be given.

With that said I can try and see if I can write a simple local wrapper function for `torch_np.meshgrid` to get the Tensor out. Correct me if I'm wrong @ezyang @yanboliang

Reply:
The way we designed `torch_np`, and how I see this is as follows:
- `torch_np` provides a class `torch_np.ndarray` that can be created from an `np.ndarray`
  - Internally, this class is a wrapper around a `torch.Tensor` that effectively implements the API from `np.ndarray` via `torch.` functions and `torch.Tensor` methods.
- The other thing that `torch_np` is implementations for all the free functions (and a few other funny objects) from the NumPy API in terms of free functions within the `torch` namespace.
- All these functions and methods accept anything that NumPy accepts (`np.ndarray`s, Python scalars, NumPy scalars, list of lists... anything that smells like a `np.ndarray` really), np.dtypes, and other NumPy-lang objects, wraps these into `torch_np.ndarray`, `torch.dtype`s and equiv. Torch-lang objects, calls one or more PyTorch operations, and returns one (or more) `torch_np.ndarray`s.

To summarise, `torch_np` functions are functions from NumPy-objects + `torch_np.ndarray`s into `torch_np.ndarray`s implemented via PyTorch functions. These functions return `torch_np.ndarray`s so that we can compose calls to their methods.

With this tooling, the way I figured this would be implemented is:
- If we see a call to an `np.foo`, dispatch it to `torch_np.foo`. Don't worry about wrapping the `np.arrays` or `np.dtypes` here, `torch_np` functions will take care fo that.
- If we see a call to an `np.ndarray` method or property or anything, wrap the object it's being called on via `torch_np.asarray` and call the same method/property on the resulting `torch_np.ndarray`.
  - After doing this, if you keep tracing, you know that all you're going to see here is good old calls to functions and methods within the torch namespace.
- Finally, when you are leaving the frame, you'll still need to convert back the `torch_np.ndarray`s into `np.array`s but that should be it.

This being said, let's see what Ed and Yanbo have to say.

Reply:
Yeah, storing a torch_np.ndarray inside the NumpyVariable seems nice and convenient. I think the proxy stuff will work out too?

Reply:
OK I did some plumbing and it seems to be working. Now `NumpyTensorVariable` is storing a `torch_np.ndarray`. The only issue I'm having is there's no way to `LOAD_ATTR` if I need to return an actual numpy ndarray.

For example:

```
def f(x: torch.Tensor):
    a = x.numpy()
    return a.T
```

Needs to be translated into something like:
```
def f(x: torch.Tensor):
    b = __compiled_function(x)
    return b.tensor.numpy()
```
Currently I'm adding some instructions for `.tensor` and `.numpy()` but for `.tensor` it failed because `tensor` doesn't exist in `code.co_names`. The workaround I can think of, is to add a method API `.to_numpy()` to `torch_np.ndarray`. Thoughts?

Reply:
> there's no way to LOAD_ATTR if I need to return an actual numpy ndarray

This is resolved. Thanks to @yanboliang